dev/test/bulk-writer.ts

```diff
@@ -0,0 +1,502 @@
+
+    const doc = firestore.doc('collectionId/doc');
+    bulkWriter.set(doc, {foo: 'bar'}).catch(err => {
+      incrementOpCount();
+      expect(err.code).to.equal(Status.UNAVAILABLE);
+    });
+
+    return bulkWriter.close().then(async () => verifyOpCount(1));
+  });
+
+  it('flush() resolves immediately if there are no writes', async () => {
+    await instantiateInstance([]);
+    return bulkWriter.flush().then(() => verifyOpCount(0));
+  });
+
+  it('adds writes to a new batch after calling flush()', async () => {
+    await instantiateInstance([
+      {
+        request: createOp('doc', 'bar'),
+        response: createResponse([successResponse(2)]),
+      },
+      {
+        request: setOp('doc2', 'bar1'),
+        response: createResponse([successResponse(2)]),
+      },
+    ]);
+    bulkWriter
+      .create(firestore.doc('collectionId/doc'), {foo: 'bar'})
+      .then(incrementOpCount);
+    bulkWriter.flush();
+    bulkWriter
+      .set(firestore.doc('collectionId/doc2'), {foo: 'bar1'})
+      .then(incrementOpCount);
+    await bulkWriter.close().then(async () => {
+      verifyOpCount(2);
+    });
+  });
+
+  it('close() sends all writes', async () => {
+    await instantiateInstance([
+      {
+        request: createOp('doc', 'bar'),
+        response: createResponse([successResponse(2)]),
+      },
+    ]);
+    const doc = firestore.doc('collectionId/doc');
+    bulkWriter.create(doc, {foo: 'bar'}).then(incrementOpCount);
+    return bulkWriter.close().then(async () => {
+      verifyOpCount(1);
+    });
+  });
+
+  it('close() resolves immediately if there are no writes', async () => {
+    await instantiateInstance([]);
+    return bulkWriter.close().then(() => verifyOpCount(0));
+  });
+
+  it('cannot call methods after close() is called', async () => {
+    await instantiateInstance([]);
+
+    const expected = 'BulkWriter has already been closed.';
+    const doc = firestore.doc('collectionId/doc');
+    await bulkWriter.close();
+    expect(() => bulkWriter.set(doc, {})).to.throw(expected);
+    expect(() => bulkWriter.create(doc, {})).to.throw(expected);
+    expect(() => bulkWriter.update(doc, {})).to.throw(expected);
+    expect(() => bulkWriter.delete(doc)).to.throw(expected);
+    expect(bulkWriter.flush()).to.eventually.be.rejectedWith(expected);
+    expect(bulkWriter.close()).to.eventually.be.rejectedWith(expected);
+  });
+
+  it('sends writes to the same document in separate batches', async () => {
+    await instantiateInstance([
+      {
+        request: setOp('doc', 'bar'),
+        response: createResponse([successResponse(0)]),
+      },
+      {
+        request: updateOp('doc', 'bar1'),
+        response: createResponse([successResponse(1)]),
+      },
+    ]);
+
+    const doc = firestore.doc('collectionId/doc');
+    bulkWriter.set(doc, {foo: 'bar'}).then(incrementOpCount);
+    bulkWriter.update(doc, {foo: 'bar1'}).then(incrementOpCount);
+
+    return bulkWriter.close().then(async () => {
```

Comment:
This test would actually have passed with your original logic that was based on `Set<DocumentReference>`. To test that your are correctly comparing references, you can change this to use with two different document references that point to the same document.